File: GaiaSerialIOClient/SerialClient.cs
using System;
using System.Text;
using StackExchange.Redis;

namespace Gaia.SerialIO
{
    /// <summary>
    /// Client for serial port input and output.
    /// </summary>
    public class SerialClient
    {
        /// <summary>
        /// Channel connection to the Redis server.
        /// </summary>
        private readonly ISubscriber Communicator;
        /// <summary>
        /// Bound device name.
        /// </summary>
        private readonly string DeviceName;
        
        /// <summary>
        /// Try to connect to the remote device.
        /// </summary>
        /// <param name="device_name"></param>
        /// <param name="port"></param>
        /// <param name="ip"></param>
        public SerialClient(string device_name, uint port = 6379, string ip = "127.0.0.1")
        {
            DeviceName = device_name;
            var connection = ConnectionMultiplexer.Connect($"{ip}:{port.ToString()}");
            Communicator = connection.GetSubscriber();

            var database = connection.GetDatabase();
            if (!database.SetContains("serial_ports", DeviceName))
            {
                throw new Exception($"Target serial port device {DeviceName} is unavailable.");
            }

            Communicator.Subscribe($"serial_ports/{DeviceName}/read", (channel, value) =>
            {
                if (OnReceive != null)
                {
                    OnReceive(Encoding.ASCII.GetBytes(value.ToString()));
                }
            });
        }
        
        /// <summary>
        /// Delegate of bytes data handler.
        /// </summary>
        public delegate void BytesHandler(byte[] data);

        /// <summary>
        /// Receive event, triggered when receive data from serial ports.
        /// </summary>
        public event BytesHandler OnReceive;

        /// <summary>
        /// Send a string text to the serial port.
        /// </summary>
        /// <param name="text"></param>
        public void Send(string text)
        {
            Communicator.Publish($"serial_ports/{DeviceName}/write", text);
        }
        
        /// <summary>
        /// Send bytes to the serial port.
        /// </summary>
        /// <param name="bytes"></param>
        public void Send(byte[] bytes)
        {
            Communicator.Publish($"serial_ports/{DeviceName}/write", 
                Encoding.ASCII.GetString(bytes));
        }

        /// <summary>
        /// Send a command text to the command channel of the serial device router.
        /// </summary>
        /// <param name="command">Command text.</param>
        public void SendCommand(string command)
        {
            Communicator.Publish($"serial_ports/{DeviceName}/command", command);
        }
    }
}

File: Native/GaiaSerialIOClient/SerialClient.cpp
#include "SerialClient.hpp"

#include <thread>

namespace Gaia::SerialIO
{
    /// Make a new connection to the Redis server.
    SerialClient::SerialClient(const std::string& device_name, unsigned int port, const std::string &ip) :
        SerialClient(device_name,
                     std::make_shared<sw::redis::Redis>("tcp://" + ip + ":" + std::to_string(port)))
    {}

    /// Reuse the connection to a Redis server.
    SerialClient::SerialClient(const std::string &device_name, std::shared_ptr<sw::redis::Redis> connection) :
            Connection(std::move(connection)),
            ChannelReadName("serial_ports/" + device_name + "/read"),
            ChannelWriteName("serial_ports/" + device_name + "/write"),
            ChannelCommandName("serial_ports/" + device_name + "/command"),
            BackgroundListenIntervalTime(std::chrono::milliseconds(10)),
            BackgroundListener([this](const auto& life_flag){
                while (life_flag)
                {
                    this->Listen();
                    std::this_thread::sleep_for(this->BackgroundListenIntervalTime.load());
                }
            })
    {
        Subscriber = std::make_unique<sw::redis::Subscriber>(Connection->subscriber());
        Subscriber->subscribe(ChannelReadName);
        Subscriber->on_message([this](const std::string& channel, std::string value)
                               {
                                   this->OnReceive.Trigger(std::move(value));
                               });
    }

    /// Send a text to the serial port.
    void SerialClient::Send(const std::string& text)
    {
        Connection->publish(ChannelWriteName, text);
    }

    /// Send a command to the command channel of the serial port.
    void SerialClient::SendCommand(const std::string& command)
    {
        Connection->publish(ChannelCommandName, command);
    }

    /// Consume serial port messages for once.
    void SerialClient::Listen()
    {
        if (Subscriber)
        {
            Subscriber->consume();
        }
    }
}